kv/dist_sender_test.go

```diff
@@ -752,3 +752,99 @@ func TestGetNodeDescriptor(t *testing.T) {
 }
+
+// TestSpanRangeScanWhenMerging simulate a range merge operation happend between get range desc
+// and get next range desc in getDescriptors function. the second sub scan use next range's start key
+// as start key of scan, this cause the the whole scan get duplicate key-values.
+func TestSpanRangeScanWhenMerging(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	g, s := makeTestGossip(t)
+	defer s()
+
+	// Asume we have two range [a-b) [b-KeyMax)
+	merged := false
+
+	var firstRange = proto.RangeDescriptor{
+		RaftID:   1,
+		StartKey: proto.Key("a"),
+		EndKey:   proto.Key("b"),
+		Replicas: []proto.Replica{
+			{
+				NodeID:  1,
+				StoreID: 1,
+			},
+		},
+	}
+
+	// Two ranges will be merged to one range
+	var mergedRange = proto.RangeDescriptor{
+		RaftID:   1,
+		StartKey: proto.Key("a"),
+		EndKey:   proto.KeyMax,
+		Replicas: []proto.Replica{
+			{
+				NodeID:  1,
+				StoreID: 1,
+			},
+		},
+	}
+	// Asume we have two key-values: a-1  c-2
+	var testFn rpcSendFn = func(_ rpc.Options, method string, addrs []net.Addr, getArgs func(addr net.Addr) interface{}, getReply func() interface{}, _ *rpc.Context) ([]interface{}, error) {
+		header := getArgs(testAddress).(proto.Request).Header()
+		reply := getReply().(*proto.ScanResponse)
+		if method == "Node.Scan" && header.Key.Equal(proto.Key("a")) && header.EndKey.Equal(proto.Key("b")) {
+
+			reply.Rows = []proto.KeyValue{
+				{Key: proto.Key("a"), Value: proto.Value{Bytes: []byte("1")}},
+			}
+			return []interface{}{reply}, nil
+		}
+
+		if method == "Node.Scan" && header.Key.Equal(proto.Key("a")) && header.EndKey.Equal(proto.Key("d")) {
+
+			reply.Rows = []proto.KeyValue{
+				{Key: proto.Key("a"), Value: proto.Value{Bytes: []byte("1")}},
+				{Key: proto.Key("c"), Value: proto.Value{Bytes: []byte("2")}},
+			}
+			return []interface{}{reply}, nil
+		}
+		if method == "Node.Scan" && header.Key.Equal(proto.Key("b")) && header.EndKey.Equal(proto.Key("d")) {
+
+			reply.Rows = []proto.KeyValue{
+				{Key: proto.Key("c"), Value: proto.Value{Bytes: []byte("2")}},
+			}
+			return []interface{}{reply}, nil
+		}
+
+		t.Fatalf("unknow method, method=%s merged=%v, startkey=%v endkey=%v ", method, merged, header.Key, header.EndKey)
+		return []interface{}{getReply()}, nil
+
+	}
+
+	ctx := &DistSenderContext{
+		rpcSend: testFn,
+		rangeDescriptorDB: mockRangeDescriptorDB(func(key proto.Key, _ lookupOptions) ([]proto.RangeDescriptor, error) {
+
+			if !merged {
+				// Asume a range merge operation happened
+				merged = true
+				return []proto.RangeDescriptor{firstRange}, nil
+			}
+
+			return []proto.RangeDescriptor{mergedRange}, nil
+		}),
+	}
+	ds := NewDistSender(ctx, g)
+
+	call := proto.ScanCall(proto.Key("a"), proto.Key("d"), 10)
+	call.Args.Header().Txn = &proto.Transaction{}
+	reply := call.Reply.(*proto.ScanResponse)
+	ds.Send(context.Background(), call)
+	if err := reply.GoError(); err != nil {
+		t.Fatalf("scan encountered error: %s", err)
+	}
+
+	if len(reply.Rows) != 2 {
```

Comment:
if you follow my `actualKV` suggestion above, here you can easily check that the result is correct by using `reflect.DeepEqual(expectedKVs, reply.Rows)`, which I would prefer since your test likely passes if the DistSender decides to just scan the first stale descriptor's range twice.